clients/src/main/java/org/apache/kafka/common/requests/AddOffsetsToTxnRequest.java

```diff
@@ -53,7 +53,7 @@ public AddOffsetsToTxnRequest(AddOffsetsToTxnRequestData data, short version) {
     }
 
     @Override
-    protected AddOffsetsToTxnRequestData data() {
+    public AddOffsetsToTxnRequestData data() {
```

Comment:
It seems like there are many classes where direct access to `data` is not actually needed. How do you feel about having a public method in `RequestUtils` that exposes `data` for the raft layer instead? Something like:

```java
public static ApiMessage requestData(AbstractRequest req)
public static ApiMessage responseData(AbstractRequest req)
```

Or is it not worth it?

Reply:
Yeah, not sure it's worth it, though I don't feel strongly. I think ultimately we're going to start relying more on the generated classes to avoid unnecessary conversions. We're now entering "phase 2" of the request overhaul which means we can start figuring out how to remove the AbstractRequest/AbstractResponse layer. I think it will take more smarts in the generated classes to make a dent here, but if we are agreed on the goal (?), then I do not think preserving the encapsulation here is worthwhile.

Reply:
I don't quite understand how we would handle versioning well if we only have data classes. Do you have thoughts on that?

Reply:
Support for optional fields would go a long way I think. I am not sure it will be possible to remove all intermediate representations, but perhaps it can be the exception and not the rule. Some version checks in `KafkaApis` are probably inevitable.

Reply:
I can see this:

1. AbstractRequest/AbstractResponse methods become part of the ApiMessage hierarchy.
2. FooRequest/FooResponse extends FooDataRequest/FooDataResponse (like Colin suggested before)

But I don't think you want to eliminate FooRequest/FooResponse in the example above. You don't need to perform conversions for the inner classes, but it's a place where you can normalize the representation. We do that for many of the existing request/response classes.